packages/react-reconciler/src/ReactChildFiber.js

```diff
@@ -1128,6 +1149,32 @@ function ChildReconciler(shouldTrackSideEffects) {
       child = child.sibling;
     }
 
+    if (
+      typeof element.type === 'object' &&
+      element.type !== null &&
+      element.type.$$typeof === REACT_FORWARD_REF_TYPE
+    ) {
+      const renderFn = element.type.renderFn;
+      invariant(
+        typeof renderFn === 'function',
+        'forwardRef requires a render function but was given %s.%s',
+        renderFn === null ? 'null' : typeof renderFn,
+        ReactDebugCurrentFiber.getCurrentFiberStackAddendum() || '',
+      );
+
+      const newElement = renderFn(element.props, element.ref);
+      if (newElement == null) {
+        return null;
+      }
+
+      return reconcileSingleElement(
```

Comment:
We've tried pretty hard to avoid recursive algorithms so that we could try to inline these into hot loops instead. Would prefer a while loop.